Automattic/jetpack
# [CRM] Add API v4 base logic

## Summary

The aim of this PR is to create a new place for our API v4 to exist, but also implement base functionality we need to start developing the same way like documentation, base controllers and test resources.

We try to achieve this by implementing a simple version of `[GET] /contacts/:id` so we have something concrete to compare to.

Fixes https://github.com/Automattic/zero-bs-crm/issues/3066

## Proposed changes:


**Feature(s)**

High level then we're adding a `[GET] /contacts/:id` endpoint since it's one of the simplest endpoints we can implement and it will serve as an example and proof of concept that our base logic is functioning.

* Add a new feature flag for experimental API v4 endpoints that we want to merge into trunk but don't want to officially ship yet (`jetpack_crm_feature_flag_api_v4`).
* Add API v4 documentation to explain the principles we follow.
* Add `REST_Base_Controller` class: Currently it "just" defines the namespace for our endpoints, but will probably grow in the future when we start implementing more endpoints and we identify generic logic.
* Add `REST_Base_Objects_Controller`: This class will serve as the base class for any DAL object endpoints since we'll have a lot of generic logic that can be applied across objects.
  * E.g.: Fetch the DAL service, but also convert object structure/models to endpoint args when we start to create create/update endpoints.
* Add `REST_Contacts_Controller`: An example of how we're going to work with API v4 moving forward and serves as the example that our API namespace + test logic works as expected.

**Tests**

High level then we're adding a REST API test suite that has the base functionality/bootrapping sorted out to be able to test future CRM API v4 endpoints:

* Add `REST_Base_Test_Case` that our endpoints can extend in order to test CRM endpoints.
* Add `REST_Authentication_Test` to provide easy ways to ensure that all of our custom endpoints fail authentication if we do not have a valid current_user().
  * To make it easier, then it should be enough to just provide new URLs to the data provider(s).
* Add `REST_Contacts_Controller_Test` to verify that our `[GET] /contacts/:id` endpoint works as expected.

### Other information:


## Jetpack product discussion



We don't have any direct discussion for implementing an endpoint that returns a contact, but it's prep work for the upcoming automations refactor: pbhBOz-3ss-p2

## Does this pull request change what data or activity we track or use?



No.

## Testing instructions:






## Automated testing

Verify that the new `rest-api` test suite doesn't throw any errors.

This can either be done by going through the CI steps on this PR or by following the below instructions on your local environment.

* Checkout branch locally
* Run: `jetpack install plugins/crm`
* Go to the CRM directory: `projects/plugins/crm`
* Run `composer run-script phpunit`

## Manual testing

* Add snippet below to your local environment (e.g. in `0-sandbox.php`).
* Update the `$contact_id` variable to a contact ID that exists in your local environment.
* Login to the website (the endpoint requires a logged in administrator user).
* Go to any front-end page and either:
  * Inspect the Network for the XHR request.
  * Inspect console since we `console.log()` the response in the snippet.
* Verify that we return the contact that corresponds to the ID.

```php
// Enable the feature flag that initialise the in progress code.
add_filter( 'jetpack_crm_feature_flag_api_v4', '__return_true' );

// Output a JS script to `wp_footer` that includes a nonce value so the WP REST registers you as a logged in user.
add_action( 'wp_footer', function () {
	$nonce = wp_create_nonce( 'wp_rest' );
	$contact_id = 1;

	$code = "
fetch( '/wp-json/jetpack-crm/v4/contacts/$contact_id', { headers: { 'X-WP-Nonce': '$nonce' } } ).then( function ( response ) {
	return response.json();
} ).then( function( data ) {
	console.log( data );
} ).catch( function ( err ) {
	console.log( err );
} );";

	printf( '<script>%s</script>', $code );
} );

```
